File: src/modules/proxy/pkg/fnapi/request.go
package fnapi

import (
	"crypto/md5"
	"encoding/json"
	"fmt"
	"io"
	"math/rand"
	"net/http"
	"strconv"
	"strings"
	"time"

	"proxy/pkg/logger"
)

const (
	apiKey    = "NDzZTVxnRKP8Z0jXg1VAMonaG8akvh"
	apiSecret = "16CCEB3D-AB42-077D-36A1-F355324E4237"
)

// ApiResponse API响应
type ApiResponse[T any] struct {
	Success          bool   `json:"success"`
	Data             T      `json:"data,omitempty"`
	Message          string `json:"message,omitempty"`
	CertificateError bool   `json:"certificateError,omitempty"`
}

// FnApiResponseData 飞牛API响应数据
type FnApiResponseData[T any] struct {
	Code int    `json:"code"`
	Msg  string `json:"msg"`
	Data T      `json:"data"`
}

// HttpMethod HTTP方法
type HttpMethod string

const (
	MethodGET    HttpMethod = "GET"
	MethodPOST   HttpMethod = "POST"
	MethodPUT    HttpMethod = "PUT"
	MethodDELETE HttpMethod = "DELETE"
)

// GetMd5 计算MD5哈希
func GetMd5(text string) string {
	hash := md5.Sum([]byte(text))
	return fmt.Sprintf("%x", hash)
}

// GenerateRandomDigits 生成随机数字字符串
func GenerateRandomDigits(start, end int) string {
	if start >= end {
		return strconv.Itoa(start)
	}
	return strconv.Itoa(rand.Intn(end-start) + start)
}

// GenFnAuthx 生成授权签名
func GenFnAuthx(url string, data interface{}) string {
	nonce := GenerateRandomDigits(100000, 1000000)
	timestamp := time.Now().Unix()
	var dataJson string
	if data != nil {
		dataBytes, _ := json.Marshal(data)
		dataJson = string(dataBytes)
	}
	dataJsonMd5 := GetMd5(dataJson)

	signArray := []string{
		apiKey,
		url,
		nonce,
		strconv.FormatInt(timestamp, 10),
		dataJsonMd5,
		apiSecret,
	}
	signStr := strings.Join(signArray, "_")
	finalSign := GetMd5(signStr)

	return fmt.Sprintf("nonce=%s&timestamp=%d&sign=%s", nonce, timestamp, finalSign)
}

const DefaultTimeout = 10000 // 默认超时时间（毫秒）

func ComposeCookieHeader(accessCookie string) string {
	parts := make([]string, 0)
	seen := make(map[string]struct{})
	for _, rawPart := range strings.Split(accessCookie, ";") {
		part := strings.TrimSpace(rawPart)
		separator := strings.Index(part, "=")
		if separator <= 0 {
			continue
		}
		name := strings.TrimSpace(part[:separator])
		normalizedName := strings.ToLower(name)
		if normalizedName == "mode" {
			continue
		}
		if _, exists := seen[normalizedName]; exists {
			continue
		}
		seen[normalizedName] = struct{}{}
		parts = append(parts, name+"="+strings.TrimSpace(part[separator+1:]))
	}
	parts = append(parts, "mode=relay")
	return strings.Join(parts, "; ")
}

// Request 发送API请求
func Request[T any](client *http.Client, baseURL, url string, method HttpMethod, token string, data interface{}, extraHeaders map[string]string, timeout, tryTimes int) (*ApiResponse[T], error) {
	if tryTimes <= 0 {
		tryTimes = 5
	}
	if timeout <= 0 {
		timeout = DefaultTimeout
	}

	fullURL := baseURL + url

	// 为POST/PUT请求添加随机数防重放
	if method == MethodPOST || method == MethodPUT {
		if data == nil {
			data = map[string]interface{}{}
		}
		// 这里假设data是map，可以添加nonce
		if m, ok := data.(map[string]interface{}); ok {
			nonce := GenerateRandomDigits(100000, 1000000)
			m["nonce"] = nonce
			logger.Debugf("为%s请求添加nonce: %s", string(method), nonce)
		}
	}

	authx := GenFnAuthx(url, data)

	headers := map[string]string{
		"Content-Type":  "application/json",
		"Authorization": token,
		"Authx":         authx,
		"Cookie":        "mode=relay",
	}
	for k, v := range extraHeaders {
		headers[k] = v
	}
	headers["Cookie"] = ComposeCookieHeader(headers["Cookie"])

	var lastErr error
	for attempt := 0; attempt <= tryTimes; attempt++ {
		logger.Debugf("尝试第%d次请求: %s", attempt+1, fullURL)

		var req *http.Request
		var err error

		var body io.Reader
		if data != nil {
			dataBytes, _ := json.Marshal(data)
			body = strings.NewReader(string(dataBytes))
			logger.Debugf("请求体大小: %d bytes", len(dataBytes))
		}

		req, err = http.NewRequest(string(method), fullURL, body)
		if err != nil {
			logger.Errorf("创建HTTP请求失败: %v", err)
			return nil, err
		}

		for k, v := range headers {
			req.Header.Set(k, v)
		}

		resp, err := client.Do(req)
		if err != nil {
			logger.Warnf("HTTP请求失败 (尝试%d/%d): %v", attempt+1, tryTimes+1, err)
			lastErr = err
			time.Sleep(100 * time.Millisecond)
			continue
		}
		respBody, err := io.ReadAll(resp.Body)
		closeErr := resp.Body.Close()
		if err != nil {
			logger.Warnf("读取响应体失败 (尝试%d/%d): %v", attempt+1, tryTimes+1, err)
			lastErr = err
			time.Sleep(100 * time.Millisecond)
			continue
		}
		if closeErr != nil {
			logger.Warnf("关闭响应体失败 (尝试%d/%d): %v", attempt+1, tryTimes+1, closeErr)
		}

		logger.Debugf("收到响应: 状态码=%d, 响应体大小=%d bytes", resp.StatusCode, len(respBody))

		var fnResp FnApiResponseData[T]
		if err := json.Unmarshal(respBody, &fnResp); err != nil {
			logger.Warnf("解析JSON响应失败 (尝试%d/%d): %v", attempt+1, tryTimes+1, err)
			lastErr = err
			time.Sleep(100 * time.Millisecond)
			continue
		}

		// 处理签名错误的重试逻辑
		if fnResp.Code == 5000 && fnResp.Msg == "invalid sign" {
			logger.Warnf("签名错误，重试中 (尝试%d/%d): %s", attempt+1, tryTimes+1, fnResp.Msg)
			if attempt >= tryTimes {
				logger.Errorf("签名错误重试次数过多")
				return &ApiResponse[T]{
					Success: false,
					Message: fmt.Sprintf("尝试次数过多 try_times = %d", attempt+1),
				}, nil
			}
			time.Sleep(100 * time.Millisecond)
			continue
		}

		// 处理业务错误
		if fnResp.Code != 0 {
			logger.Errorf("API业务错误: 代码=%d, 消息=%s", fnResp.Code, fnResp.Msg)
			return &ApiResponse[T]{
				Success: false,
				Message: fnResp.Msg,
			}, nil
		}

		logger.Infof("API请求成功: %s %s", string(method), url)
		return &ApiResponse[T]{
			Success: true,
			Data:    fnResp.Data,
		}, nil
	}

	logger.Errorf("API请求最终失败: %v", lastErr)
	return &ApiResponse[T]{
		Success: false,
		Message: fmt.Sprintf("请求失败: %v", lastErr),
	}, lastErr
}

File: src/modules/fn_api/request.ts
import * as crypto from 'crypto';
import axios, { AxiosResponse } from 'axios';
import { setTimeout } from 'timers/promises';
import https from 'https';
import log from '../logger';
import { isTrusted, showCertificateTrustDialog, isCertificateError, addTrustedHost } from '../cert_trust';
import { composeCookieHeader, getAccessCookieHeader } from './accessGrant';

// 全局配置
const api_key = 'NDzZTVxnRKP8Z0jXg1VAMonaG8akvh';
const api_secret = '16CCEB3D-AB42-077D-36A1-F355324E4237';

// 类型定义
export interface ApiResponse<T = any> {
    success: boolean;
    data?: T;
    message?: string;
    certificateError?: boolean; // 标识是否为证书错误
    moveUrl?: string; // 重定向URL
}

export interface FnApiResponseData<T = any> {
    code: number;
    msg: string;
    data: T;
}

export enum HttpMethod {
    GET = 'get',
    POST = 'post',
    PUT = 'put',
    DELETE = 'delete'
}

// MD5哈希计算
export function getMd5(text: string): string {
    return crypto.createHash('md5').update(text, 'utf8').digest('hex');
}

// 生成随机数字字符串
export function generateRandomDigits(start: number = 100000, end: number = 1000000): string {
    return Math.floor(Math.random() * (end - start) + start).toString();
}

// 生成授权签名
export function genFnAuthx(url: string, data?: any): string {
    const nonce = generateRandomDigits();
    const timestamp = Date.now();
    const dataJson = data ? JSON.stringify(data) : '';
    const dataJsonMd5 = getMd5(dataJson);

    const signArray = [
        api_key,
        url,
        nonce,
        timestamp.toString(),
        dataJsonMd5,
        api_secret
    ];

    const signStr = signArray.join('_');
    return `nonce=${nonce}&timestamp=${timestamp}&sign=${getMd5(signStr)}`;
}

// 默认超时时间（毫秒）
export const DEFAULT_TIMEOUT = 10000;

// API请求函数
export async function request<T = any>(
    baseUrl: string,
    url: string,
    method: HttpMethod,
    token: string,
    data?: any,
    extraHeaders?: Record<string, string>,
    axiosConfig?: Record<string, any>, // 外部传入的 axios 配置
    timeout: number = DEFAULT_TIMEOUT,
    tryTimes: number = 5,
): Promise<ApiResponse<T>> {
    const normalizedBaseUrl = baseUrl.replace(/\/+$/, '');
    const fullUrl = normalizedBaseUrl + url;
    if (method === HttpMethod.POST || method === HttpMethod.PUT) {
        data = data || {};
        data["nonce"] = generateRandomDigits(); // POST/PUT请求添加随机数防重放
    }

    const authx = genFnAuthx(url, data);

    const headers = {
        "Content-Type": "application/json",
        "Authorization": token,
        "Authx": authx,
        ...extraHeaders,
        "Cookie": composeCookieHeader(getAccessCookieHeader(normalizedBaseUrl), extraHeaders?.Cookie),
    };

    // 根据URL是否已被信任来决定是否验证证书
    const shouldIgnoreCert = isTrusted(normalizedBaseUrl);

    const config = {
        headers,
        timeout: timeout,
        // 禁止 Axios 自动处理重定向，手动处理
        maxRedirects: 0,
        // 允许 3xx 状态码进入 .then() 而不是 .catch()
        validateStatus: (status: number) => status >= 200 && status < 400,
        httpsAgent: new https.Agent({
            rejectUnauthorized: !shouldIgnoreCert,
            keepAlive: true,
            timeout: timeout,
            maxSockets: 10,
        }),
        ...axiosConfig
    };

    for (let attempt = 0; attempt <= tryTimes; attempt++) {
        try {
            let response: AxiosResponse<FnApiResponseData<T>>;

            switch (method) {
                case HttpMethod.GET: response = await axios.get(fullUrl, config); break;
                case HttpMethod.POST: response = await axios.post(fullUrl, data, config); break;
                case HttpMethod.PUT: response = await axios.put(fullUrl, data, config); break;
                case HttpMethod.DELETE: response = await axios.delete(fullUrl, config); break;
                default: throw new Error(`Unsupported method: ${method}`);
            }

            if ([301, 302, 307, 308].includes(response.status)) {
                const location = response.headers.location;
                log.warn(`检测到重定向 (${response.status}) -> ${location}`);

                if (location) {
                    // 1. 解析新地址
                    let newBaseUrl = normalizedBaseUrl;
                    let newUrlPath = location;

                    // 如果是绝对路径 (http开头)，重新拆解 baseUrl 和 path
                    if (location.startsWith('http')) {
                        const parsedUrl = new URL(location);
                        newBaseUrl = `${parsedUrl.protocol}//${parsedUrl.host}`;
                        newUrlPath = parsedUrl.pathname + parsedUrl.search;
                    }

                    const recursiveResult = await request<T>(
                        newBaseUrl,
                        newUrlPath,
                        method,
                        token,
                        data,
                        extraHeaders,
                        axiosConfig,
                        timeout,
                        tryTimes - 1
                    );

                    return {
                        ...recursiveResult,
                        moveUrl: recursiveResult.moveUrl || newBaseUrl
                    };
                }
            }

            // Axios 允许 header 值为多种类型；HTTP Content-Type 只按字符串处理。
            const contentType = response.headers['content-type'];
            if (typeof contentType === 'string' && !contentType.includes('application/json')) {
                return {
                    success: true,
                    data: response.data as any, // 直接返回原始数据
                };
            }

            const res = response.data;

            // 处理签名错误的重试逻辑
            if (res.code === 5000 && res.msg === 'invalid sign') {
                if (attempt >= tryTimes) {
                    return {
                        success: false,
                        message: `尝试次数过多 try_times = ${attempt + 1}`
                    };
                }

                log.warn(`fn_api 请求时签名错误，重试中 attempt = ${attempt + 1}, path: ${url}`);
                await setTimeout(100); // 等待100ms
                continue; // 继续下一次循环
            }

            // 处理业务错误
            if (res.code !== 0) {
                // 注意：如果重定向返回 HTML，res.code 会是 undefined，这里要小心
                log.error(`fn_api 请求失败`, res);
                return { success: false, message: res.msg || `HTTP Error ${response.status}` };
            }

            return {
                success: true,
                data: res.data
            };

        } catch (error: any) {
            const errorCode = error.code || 'UNKNOWN';
            // 优先获取 error.message，因为 connection error 没有 response
            const errorMsg = error.message;
            const responseStatus = error.response?.status ?? '无响应';

            log.error(`请求异常: [${errorCode}] ${errorMsg} | HTTP: ${responseStatus} | path: ${url}`);

            // 检查是否为证书验证错误且URL未被信任
            if (isCertificateError(error) && !isTrusted(normalizedBaseUrl)) {
                log.warn(`检测到证书验证错误: code: ${errorCode}, msg: ${errorMsg}, path: ${url}`);

                // 返回特殊的证书错误响应，让上层处理
                return {
                    success: false,
                    message: errorMsg,
                    // 添加一个特殊标识表示这是证书错误
                    certificateError: true
                } as ApiResponse<T> & { certificateError?: boolean };
            }

            // 如果是最后一次尝试，返回错误
            if (attempt >= tryTimes) {
                return { success: false, message: errorMsg };
            }

            await setTimeout(100);
        }
    }

    // 这行代码理论上不会执行到，但为了类型安全加上
    return {
        success: false,
        message: '重试逻辑异常'
    };
}
